### DISCUSSION TITLE
Guided Exercise - Microservices - Declarative Programming - Spring OpenFeign

### DISCUSSION THREAD

**Discussion by akash-coded** [akash-coded/spring-framework#159] [Tasks]
Design a comprehensive Spring Boot microservices that incorporates Feign, custom error handling, and other aspects.

### Scenario: E-commerce App

Imagine a simple e-commerce application where we have a `Product Service`, `Order Service`, and `User Service`.

- `Product Service`: Handles CRUD operations for products.
- `Order Service`: Handles CRUD operations for orders. Depends on the `Product Service` to verify product availability.
- `User Service`: Manages user profiles.

### Architecture

- Product Service (Port: 8081)
- Order Service (Port: 8082)
- User Service (Port: 8083)

#### Dependencies to include

For each microservice, include these dependencies in your `build.gradle` or `pom.xml`:

- Spring Web
- Spring Data JPA
- H2 Database
- Spring Cloud OpenFeign (for Order Service)

##### For Order Service, add OpenFeign:

```xml
<dependency>
    <groupId>org.springframework.cloud</groupId>
    <artifactId>spring-cloud-starter-openfeign</artifactId>
</dependency>
```

#### Application Properties for Each Microservice

**Product Service (src/main/resources/application.properties)**

```properties
server.port=8081
```

**Order Service (src/main/resources/application.properties)**

```properties
server.port=8082
feign.client.config.default.connectTimeout=5000
feign.client.config.default.readTimeout=5000
```

**User Service (src/main/resources/application.properties)**

```properties
server.port=8083
```

### Feign Configuration for Order Service

Create a custom `ErrorDecoder` as previously described.

### Code Repository Structure

```plaintext
- ecommerce-app
  - product-service
    - src
      - main
        - java
          - com
            - productservice
              - controller
              - model
              - repository
        - resources
    - build.gradle
  - order-service
    - src
      - main
        - java
          - com
            - orderservice
              - client
              - controller
              - model
              - repository
              - config
        - resources
    - build.gradle
  - user-service
    - src
      - main
        - java
          - com
            - userservice
              - controller
              - model
              - repository
        - resources
    - build.gradle
```

#### Testing the finished functionality

Here are some example API routes you could use to test:

- **Add Product**: `POST http://localhost:8081/products`
  - Payload: `{ "name": "iPhone", "price": 999.99 }`
- **Place Order**: `POST http://localhost:8082/orders`
  - Payload: `{ "userId": 1, "productId": 1 }`
- **Get User Profile**: `GET http://localhost:8083/users/1`

You would expect:

- For Add Product: Successful addition of product and returned product information.
- For Place Order: Verification of product existence via Feign to Product Service and then successful order creation.
- For Get User Profile: User profile information.

#### Do's and Don'ts:

- **Do**: Keep your Feign clients and the rest of the code decoupled. Use a service to bridge them.
- **Don't**: Hardcode URLs or rely on specific HTTP status codes.
- **Do**: Use Feign's built-in logging and metrics capture capabilities to monitor the behavior of your remote calls.
  
#### Optimization:

- Adjust the Feign timeouts and retry mechanisms according to your SLAs.
- Use connection pools for Feign clients.

### Rationalization & Internal Details

The `Order Service` needs to talk to the `Product Service` to verify the availability of products. This is a classic use-case for using Spring Cloud Feign. Feign simplifies writing HTTP clients. The error decoder allows centralized control over exception handling.

**Note**: This exercise should give you a holistic view of how to design your microservices with Spring Cloud Feign.

---

Below is a further detailed breakdown of each microservice, focusing on code snippets and how they interact.

#### Model

```java
@Entity
public class Product {
    @Id
    @GeneratedValue(strategy = GenerationType.AUTO)
    private Long id;
    private String name;
    private double price;
    // Getter and Setters
}
```

#### Repository

```java
public interface ProductRepository extends JpaRepository<Product, Long> {
}
```

#### Controller

```java
@RestController
@RequestMapping("/products")
public class ProductController {
    @Autowired
    private ProductRepository productRepository;

    @PostMapping
    public Product createProduct(@RequestBody Product product) {
        return productRepository.save(product);
    }
}
```

### 2. User Service (Simplified)

Since the User service is not central to demonstrating Feign, we can assume it provides basic user profile functionality.

#### Model

```java
@Entity
public class Order {
    @Id
    @GeneratedValue(strategy = GenerationType.AUTO)
    private Long id;
    private Long userId;
    private Long productId;
    // Getter and Setters
}
```

#### Repository

```java
public interface OrderRepository extends JpaRepository<Order, Long> {
}
```

#### Feign Client to talk to Product Service

```java
@FeignClient(name = "product-service", url = "localhost:8081")
public interface ProductClient {
    @GetMapping("/products/{id}")
    Product getProductById(@PathVariable Long id);
}
```

#### Controller

```java
@RestController
@RequestMapping("/orders")
public class OrderController {
    @Autowired
    private OrderRepository orderRepository;

    @Autowired
    private ProductClient productClient;

    @PostMapping
    public Order createOrder(@RequestBody Order order) {
        Product product = productClient.getProductById(order.getProductId());
        if (product != null) {
            return orderRepository.save(order);
        } else {
            throw new ProductNotFoundException();
        }
    }
}
```

### Error Decoder for Feign Client

```java
public class CustomErrorDecoder implements ErrorDecoder {
    @Override
    public Exception decode(String methodKey, Response response) {
        switch (response.status()) {
            case 400:
                return new BadRequestException();
            case 404:
                return new NotFoundException();
            case 500:
                return new InternalServerError();
            default:
                return new Exception("Generic error");
        }
    }
}
```

### Testing

You can start each service and then try to create an order using Postman or Curl. You should see that the Feign client is automatically handling the HTTP request to the Product service. If the product does not exist, the custom error decoder will kick in.

#### Postman Endpoints

- Add Product: POST to `http://localhost:8081/products`
- Create Order: POST to `http://localhost:8082/orders`

#### CURL Example

Adding a product:

```bash
curl -X POST http://localhost:8081/products -d '{"name": "iPhone", "price": 999}' -H "Content-Type: application/json"
```

Creating an order:

```bash
curl -X POST http://localhost:8082/orders -d '{"userId": 1, "productId": 1}' -H "Content-Type: application/json"
```

### Best Practices and Rationalization

1. **Circuit Breaker**: Consider adding a circuit breaker like Hystrix to handle failures gracefully.
2. **Timeouts**: Be mindful of the timeouts configured for Feign. Too low and you'll get unnecessary failures; too high and you'll have high latency.
3. **Error Handling**: Utilize Feign's error decoder to uniformly handle errors across all services.
4. **Feign Configuration**: Consider creating a centralized Feign configuration if you have multiple Feign clients.
5. **Separate Concerns**: Keep Feign clients separate from your business logic. Use service classes as an intermediary.

By adhering to these practices, you can build a robust and maintainable microservices architecture. Feign plays a crucial role in simplifying HTTP calls between services, thus promoting maintainability and scalability.

---

### Code Repository Structure

Here is a suggested code repository structure for each microservice:

```
product-service/
|-- src/
|   |-- main/
|   |   |-- java/
|   |   |   |-- controller/
|   |   |   |-- model/
|   |   |   |-- repository/
|   |   |-- resources/
|   |       |-- application.properties
|-- pom.xml

user-service/
|-- src/
|   |-- main/
|   |   |-- java/
|   |   |   |-- controller/
|   |   |   |-- model/
|   |   |   |-- repository/
|   |   |-- resources/
|   |       |-- application.properties
|-- pom.xml

order-service/
|-- src/
|   |-- main/
|   |   |-- java/
|   |   |   |-- controller/
|   |   |   |-- model/
|   |   |   |-- repository/
|   |   |   |-- feign/
|   |   |   |-- exception/
|   |   |-- resources/
|   |       |-- application.properties
|-- pom.xml
```

### How Microservices Will Interact

1. The `Order Service` needs product information before placing an order. It uses Feign to call `Product Service`.
2. User validations, not shown here, can be done by a similar Feign client interaction between `Order Service` and `User Service`.

### Application Properties

For `Product Service` (`application.properties`):

```properties
server.port=8081
```

For `Order Service` (`application.properties`):

```properties
server.port=8082
feign.client.config.default.connectTimeout=5000
feign.client.config.default.readTimeout=5000
```

### Spring Open Feign

Spring Open Feign makes writing HTTP clients easier through declarative templates. Using annotations, you can create HTTP Requests—such as GET, POST—and automatically encode objects into JSON.

#### Advantages

1. **Simplicity**: Avoids verbose code. No need to open connections, close them, or handle retries.
2. **Robust Error Handling**: Integrated easily with Error Decoders.
3. **Load Balancing**: Works seamlessly with Ribbon for load balancing.

#### Declarative Programming

Declarative programming is a paradigm where you express what you want to accomplish without specifying how to do it. In the context of Feign, you define a proxy interface and annotate it. Spring and Feign take care of the implementation.

### Custom Error Handling

We have already implemented a `CustomErrorDecoder` for Feign in the `Order Service`. This decodes the HTTP response status into custom exceptions.

### Best Practices, Dos, and Don'ts

1. **Do**: Use a central Feign Configuration if you have multiple Feign clients.
2. **Don't**: Put business logic in Feign Clients.
3. **Do**: Use Feign's built-in logger and logging levels for debugging.
4. **Don't**: Ignore timeouts and retry mechanisms. Configure them thoughtfully.
5. **Do**: Combine Feign with Circuit Breakers like Hystrix.

### Testing the Finished App

Use CURL or Postman to test the individual microservices, and then test them together.

1. **Add Product**: POST `http://localhost:8081/products` with JSON payload.
2. **Add User**: POST `http://localhost:8083/users` with JSON payload.
3. **Create Order**: POST `http://localhost:8082/orders` with JSON payload.

Each POST will return an object with an ID. Use these IDs to cross-verify if the services are interacting correctly.

This should provide a comprehensive view of how to build a simple yet educational set of microservices using Spring Boot and Feign.

**Comment by akash-coded** [comment]
## Deep-dive of Product Microservice Implementation

Below is a detailed walk-through of a simplified yet production-grade `Product Service` implemented using Spring Boot.

### Project Structure:

```
product-service/
|-- src/
|   |-- main/
|   |   |-- java/
|   |   |   |-- com/
|   |   |   |   |-- productservice/
|   |   |   |   |   |-- controller/
|   |   |   |   |   |   |-- ProductController.java
|   |   |   |   |   |-- model/
|   |   |   |   |   |   |-- Product.java
|   |   |   |   |   |-- repository/
|   |   |   |   |   |   |-- ProductRepository.java
|   |   |   |   |   |-- service/
|   |   |   |   |   |   |-- ProductService.java
|   |   |   |   |   |-- ProductServiceApplication.java
|   |   |-- resources/
|   |   |   |-- application.properties
|-- pom.xml
```

### POM.xml Dependencies

You would need to add these dependencies in your `pom.xml`:

```xml
<dependencies>
    <dependency>
        <groupId>org.springframework.boot</groupId>
        <artifactId>spring-boot-starter-data-jpa</artifactId>
    </dependency>
    <dependency>
        <groupId>org.springframework.boot</groupId>
        <artifactId>spring-boot-starter-web</artifactId>
    </dependency>
    <dependency>
        <groupId>com.h2database</groupId>
        <artifactId>h2</artifactId>
        <scope>runtime</scope>
    </dependency>
</dependencies>
```

### Product.java (Model)

Here we define our entity.

```java
package com.productservice.model;

import javax.persistence.Entity;
import javax.persistence.Id;

// This is our domain model
@Entity
public class Product {
    
    @Id
    private Long id;
    private String name;
    private double price;

    // Getters and setters

    // Override toString, equals and hashCode
}
```

### ProductRepository.java (Repository Interface)

This is where Spring Data JPA magic happens. Just by extending `JpaRepository`, we get a lot of CRUD functionality.

```java
package com.productservice.repository;

import com.productservice.model.Product;
import org.springframework.data.jpa.repository.JpaRepository;

public interface ProductRepository extends JpaRepository<Product, Long> {
    // We can also add custom query methods here
}
```

### ProductService.java (Service Layer)

```java
package com.productservice.service;

import com.productservice.model.Product;
import com.productservice.repository.ProductRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

@Service
public class ProductService {

    @Autowired
    private ProductRepository productRepository;

    public Product saveProduct(Product product) {
        return productRepository.save(product);
    }

    // Implement other CRUD operations and any business logic here
}
```

### ProductController.java (Controller Layer)

```java
package com.productservice.controller;

import com.productservice.model.Product;
import com.productservice.service.ProductService;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

@RestController
@RequestMapping("/products")
public class ProductController {

    @Autowired
    private ProductService productService;

    @PostMapping
    public ResponseEntity<Product> createProduct(@RequestBody Product product) {
        Product savedProduct = productService.saveProduct(product);
        return ResponseEntity.ok(savedProduct);
    }

    // Implement other CRUD APIs and possibly some APIs for querying
}
```

### Application.properties

```properties
spring.datasource.url=jdbc:h2:mem:testdb
spring.datasource.driverClassName=org.h2.Driver
spring.datasource.username=sa
spring.datasource.password=password
spring.jpa.database-platform=org.hibernate.dialect.H2Dialect
```

In this `Product Service`, we have followed best practices such as:

1. **Separation of Concerns**: Different layers for different types of logic.
2. **Database versioning and management**: H2 is used for simplicity, but in a production-grade app, you could use a more robust database.
3. **Centralized Exception Handling**: Not shown here, but you would typically use `@ControllerAdvice` to handle exceptions globally.

With these, you have a production-grade `Product Service`. Similar implementations can be followed for `User Service` and `Order Service`.

**Comment by akash-coded** [comment]
## Deep-dive of Order Microservice

Below is a deep dive into a simplified yet production-grade `Order Service` microservice implemented using Spring Boot.

### Project Structure:

```
order-service/
|-- src/
|   |-- main/
|   |   |-- java/
|   |   |   |-- com/
|   |   |   |   |-- orderservice/
|   |   |   |   |   |-- controller/
|   |   |   |   |   |   |-- OrderController.java
|   |   |   |   |   |-- model/
|   |   |   |   |   |   |-- Order.java
|   |   |   |   |   |-- repository/
|   |   |   |   |   |   |-- OrderRepository.java
|   |   |   |   |   |-- service/
|   |   |   |   |   |   |-- OrderService.java
|   |   |   |   |   |-- config/
|   |   |   |   |   |   |-- FeignConfig.java
|   |   |   |   |   |-- OrderServiceApplication.java
|   |   |-- resources/
|   |   |   |-- application.properties
|-- pom.xml
```

### POM.xml Dependencies

Add the following dependencies to your `pom.xml`:

```xml
<dependencies>
    <dependency>
        <groupId>org.springframework.boot</groupId>
        <artifactId>spring-boot-starter-data-jpa</artifactId>
    </dependency>
    <dependency>
        <groupId>org.springframework.boot</groupId>
        <artifactId>spring-boot-starter-web</artifactId>
    </dependency>
    <dependency>
        <groupId>org.springframework.cloud</groupId>
        <artifactId>spring-cloud-starter-openfeign</artifactId>
    </dependency>
    <dependency>
        <groupId>com.h2database</groupId>
        <artifactId>h2</artifactId>
        <scope>runtime</scope>
    </dependency>
</dependencies>
```

### Order.java (Model)

Here we define our entity `Order`.

```java
package com.orderservice.model;

import javax.persistence.Entity;
import javax.persistence.Id;

// The @Entity annotation marks this class as a JPA entity
@Entity
public class Order {

    @Id
    private Long id;
    private Long productId;
    private int quantity;

    // Getters and Setters, equals, hashCode, and toString
}
```

### OrderRepository.java (Repository)

```java
package com.orderservice.repository;

import com.orderservice.model.Order;
import org.springframework.data.jpa.repository.JpaRepository;

public interface OrderRepository extends JpaRepository<Order, Long> {
}
```

### OrderService.java (Service Layer)

This layer takes care of business logic.

```java
package com.orderservice.service;

import com.orderservice.model.Order;
import com.orderservice.repository.OrderRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

@Service
public class OrderService {

    @Autowired
    private OrderRepository orderRepository;

    public Order saveOrder(Order order) {
        return orderRepository.save(order);
    }

    // Implement other CRUD operations and business logic
}
```

### OrderController.java (Controller)

Here we expose the API for orders.

```java
package com.orderservice.controller;

import com.orderservice.model.Order;
import com.orderservice.service.OrderService;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

@RestController
@RequestMapping("/orders")
public class OrderController {

    @Autowired
    private OrderService orderService;

    @PostMapping
    public ResponseEntity<Order> createOrder(@RequestBody Order order) {
        Order savedOrder = orderService.saveOrder(order);
        return ResponseEntity.ok(savedOrder);
    }

    // Implement other CRUD APIs
}
```

### FeignConfig.java (Feign Configuration for Inter-Service Calls)

In a more complex microservice system, the `Order Service` might need to communicate with a `Product Service` to verify product details. This is where Feign comes in handy.

```java
package com.orderservice.config;

import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;
import feign.Logger;

@Configuration
public class FeignConfig {
    @Bean
    Logger.Level feignLoggerLevel() {
        return Logger.Level.FULL;
    }
}
```

### Application.properties

```properties
spring.datasource.url=jdbc:h2:mem:testdb
spring.datasource.driverClassName=org.h2.Driver
spring.datasource.username=sa
spring.datasource.password=password
spring.jpa.database-platform=org.hibernate.dialect.H2Dialect
```

### Points to Note:
1. **Separation of Concerns**: Different layers for different types of logic.
2. **Database Interactions**: We're using Spring Data JPA for simplifying database interactions.
3. **Inter-Service Communication**: Although not shown here, Feign could be used for calling other services.

With these elements, you have a production-grade `Order Service`. Similar best practices should be employed for other services like `User Service` and `Product Service`.

**Best Practices:**
- For production, you would replace the H2 database with a more robust database.
- Implement robust exception handling using `@ControllerAdvice`.
- Apply security measures like OAuth2 for authentication and authorization.
- Consider resilience patterns like circuit breakers for fault tolerance.
- Log meaningful information for monitoring and debugging.

This deep dive covers creating a production-grade microservice focusing on `Order Service`. Similar approaches can be taken for other services.

**Comment by akash-coded** [comment]
## Deep-dive of the User Microservice

Let's dive into the User Service and approach it from a production-grade angle.

### User Service Deep Dive

#### Code Repository Structure:

```bash
user-service
├── src
│   ├── main
│   │   ├── java
│   │   │   └── com
│   │   │       └── example
│   │   │           └── userservice
│   │   │               ├── config       # Service Configuration
│   │   │               ├── controller   # REST Controllers
│   │   │               ├── model        # Data Models
│   │   │               ├── repository   # Database Repositories
│   │   │               └── service      # Business Logic
│   │   └── resources
│   │       └── application.properties   # Application Properties
└── pom.xml                              # Project Configuration
```

#### Configuration:

In your `pom.xml`, include dependencies like JPA, H2 Database, Web, etc.

```xml
<!-- Add necessary dependencies here like JPA, Web, H2 Database -->
```

#### application.properties:

```properties
# Port and Database Configurations
server.port=8084
spring.datasource.url=jdbc:h2:mem:testdb
spring.datasource.driverClassName=org.h2.Driver
spring.datasource.username=sa
spring.datasource.password=password
spring.jpa.database-platform=org.hibernate.dialect.H2Dialect
```

#### Data Models:

```java
@Entity
public class User {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;
    private String username;
    private String email;
    
    // getters and setters
}
```

#### Repositories:

```java
public interface UserRepository extends JpaRepository<User, Long> {
    Optional<User> findByUsername(String username);
}
```

#### Business Logic:

```java
@Service
public class UserService {
    @Autowired
    private UserRepository userRepository;
    
    public User createUser(User user) {
        return userRepository.save(user);
    }
    
    public User findUserById(Long id) {
        return userRepository.findById(id).orElseThrow(() -> new UserNotFoundException("User not found"));
    }
}
```

#### Controller:

```java
@RestController
@RequestMapping("/users")
public class UserController {
    @Autowired
    private UserService userService;
    
    @PostMapping
    public ResponseEntity<User> createUser(@RequestBody User user) {
        User newUser = userService.createUser(user);
        return new ResponseEntity<>(newUser, HttpStatus.CREATED);
    }
    
    @GetMapping("/{id}")
    public ResponseEntity<User> findUserById(@PathVariable Long id) {
        User user = userService.findUserById(id);
        return new ResponseEntity<>(user, HttpStatus.OK);
    }
}
```

### Production-Grade Coding Angle:

- **Logging**: Always include logs that are helpful for debugging. For example, log the input parameters and the response or any exceptions.

```java
    public User createUser(User user) {
        log.info("Creating user with username: {}", user.getUsername());
        User createdUser = userRepository.save(user);
        log.info("User created: {}", createdUser);
        return createdUser;
    }
```

- **Exception Handling**: Use a global exception handler or specific catch blocks to return meaningful error messages to the client.

```java
@ExceptionHandler(UserNotFoundException.class)
public ResponseEntity<String> handleUserNotFound(UserNotFoundException ex) {
    return new ResponseEntity<>(ex.getMessage(), HttpStatus.NOT_FOUND);
}
```

- **Documentation**: Comment the code and possibly use tools like Swagger for API documentation.
  
- **Testing**: Add Unit Tests and Integration Tests to cover different scenarios. Tools like JUnit and Mockito can be very useful.

By following these best practices, you ensure that the User Service is robust, maintainable, and fit for a production environment.

---

## Deep-dive of the Feign-client Microservice

Let's deep-dive into a Feign client service, which acts as the client-side mechanism for communicating with other microservices. This service will be designed to talk to the `Order Service` and `Product Service`.

### Project Structure:

```plaintext
feign-client-service/
|-- src/
|   |-- main/
|   |   |-- java/
|   |   |   |-- com/
|   |   |   |   |-- feignclientservice/
|   |   |   |   |   |-- client/
|   |   |   |   |   |   |-- OrderClient.java
|   |   |   |   |   |   |-- ProductClient.java
|   |   |   |   |   |-- config/
|   |   |   |   |   |   |-- FeignErrorDecoder.java
|   |   |   |   |   |-- FeignClientServiceApplication.java
|   |   |-- resources/
|   |   |   |-- application.properties
|-- pom.xml
```

### POM.xml Dependencies

Here is the minimum required dependency for a Spring Feign client:

```xml
<dependency>
    <groupId>org.springframework.cloud</groupId>
        <artifactId>spring-cloud-starter-openfeign</artifactId>
</dependency>
```

### OrderClient.java (Feign Client for Order Service)

```java
package com.feignclientservice.client;

import com.feignclientservice.model.Order;
import org.springframework.cloud.openfeign.FeignClient;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PathVariable;

// Indicates that this interface is a Feign client with the service ID 'order-service'
@FeignClient(name = "order-service")
public interface OrderClient {

    // Maps to /orders/{id} API of the order-service
    @GetMapping("/orders/{id}")
    Order getOrderById(@PathVariable("id") Long id);

    // Implement other required APIs
}
```

### ProductClient.java (Feign Client for Product Service)

```java
package com.feignclientservice.client;

import com.feignclientservice.model.Product;
import org.springframework.cloud.openfeign.FeignClient;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PathVariable;

// Indicates that this is a Feign client with the service ID 'product-service'
@FeignClient(name = "product-service")
public interface ProductClient {

    // Maps to /products/{id} API of the product-service
    @GetMapping("/products/{id}")
    Product getProductById(@PathVariable("id") Long id);

    // Implement other required APIs
}
```

### FeignErrorDecoder.java (Feign Configuration)

This class will handle error decoding when the Feign client encounters a problem.

```java
package com.feignclientservice.config;

import feign.Response;
import feign.codec.ErrorDecoder;
import org.springframework.context.annotation.Bean;

public class FeignErrorDecoder implements ErrorDecoder {

    @Override
    public Exception decode(String methodKey, Response response) {
        switch (response.status()) {
            case 400:
                // Return custom exception
            case 404:
                // Return custom exception
            default:
                return new Exception("Generic exception");
        }
    }

    @Bean
    public ErrorDecoder errorDecoder() {
        return new FeignErrorDecoder();
    }
}
```

### FeignClientServiceApplication.java (Main Application Class)

```java
package com.feignclientservice;

import org.springframework.boot.SpringApplication;
import org.springframework.boot.autoconfigure.SpringBootApplication;
import org.springframework.cloud.openfeign.EnableFeignClients;

// Enable Feign clients
@EnableFeignClients
@SpringBootApplication
public class FeignClientServiceApplication {

    public static void main(String[] args) {
        SpringApplication.run(FeignClientServiceApplication.class, args);
    }
}
```

### Best Practices and Dos and Don'ts

1. **Exception Handling**: Always have a Feign `ErrorDecoder` to gracefully handle error scenarios.
2. **Logging**: Enable Feign logging for monitoring API calls between services.
3. **Timeouts**: Set Feign timeouts in `application.properties` to handle scenarios where the target service is slow to respond.
4. **Circuit Breaker**: Integrate Feign with Hystrix or Resilience4j for implementing the Circuit Breaker pattern.

### Application.properties

Here you can set various Feign client properties like timeouts:

```properties
feign.client.config.default.connectTimeout=5000
feign.client.config.default.readTimeout=5000
```

In this deep dive, you have seen how to set up a Feign client to communicate with other microservices in a production-grade manner. The Feign client service acts as the intermediary for invoking RESTful services in a type-safe way, abstracting the HTTP client code.

**Comment by akash-coded** [comment]
## Full workflow

Let's continue by describing how this Feign client service can be integrated with other microservices and how it functions within a full workflow. After that, we'll discuss how to test the microservices.

### Full Workflow:

1. **Order Creation**: The client application first makes a request to create an order. This request goes to the `Order Service`.
2. **Product Validation**: Before confirming the order, the `Order Service` uses the Feign client to call `Product Service` to check the availability of products.
3. **Confirmation**: Once the product's availability is confirmed, `Order Service` proceeds with the order creation and sends back a response to the client.

This workflow signifies a classic example where a Feign client is essential to enable communication between microservices.

### Test the Functionality:

#### Test `Product Service`
You can use Postman or Curl to test:

- Get product by ID: 
    - URL: `http://localhost:8081/products/1`
    - Method: GET
    - Expected Success Response: `HTTP 200 OK`
    - Expected Error Response: `HTTP 404 NOT FOUND`

#### Test `Order Service`
- Create Order:
    - URL: `http://localhost:8082/orders`
    - Method: POST
    - Payload: `{ "productId": 1, "quantity": 1 }`
    - Expected Success Response: `HTTP 201 Created`
    - Expected Error Response: `HTTP 400 BAD REQUEST`

#### Test Feign Client in Action
- Inside `Order Service`, use the Feign client to call `Product Service` and validate a product. If the product doesn't exist, you should get a `404 NOT FOUND` and the order should not be placed.

#### Application Properties:

For `FeignClientService`:
```properties
server.port=8083
```

For `ProductService`:
```properties
server.port=8081
```

For `OrderService`:
```properties
server.port=8082
```

#### Testing:
You should be able to test the above scenarios with Postman or any other HTTP client. For example:

- For creating an order which uses the `Product Service` for validation, you could hit `http://localhost:8082/orders` with POST and the required payload.
  
### Internal Details:

1. **Type Safety**: One of the main advantages of using Feign is that it provides type-safe methods to invoke RESTful services.
2. **Load Balancing**: Feign easily integrates with Ribbon to provide client-side load balancing.
3. **Error Decoder**: Custom error handling can be implemented through Feign's `ErrorDecoder`.

### Best Practices and Optimization:

- **Resiliency**: Implement Circuit Breaker and Fallback methods for your Feign client.
- **Logging and Metrics**: Enable metrics and logs to debug and monitor Feign API calls.
- **Timeouts and Retry**: Always configure sensible timeouts and retry mechanisms.

By understanding each piece of the Feign client service deeply and how it interacts with other microservices, you're in a better position to design, develop, and debug microservices architecture in a production-grade way.

**Comment by jay4tech** [comment]
Hi @akash-coded 
Name: Jay Prakash Kumar


https://github.com/jay4tech/product-service
https://github.com/jay4tech/order-service/tree/msa_159
https://github.com/jay4tech/user-service

http://localhost:8081/swagger-ui/index.html. -- Product Service
http://localhost:8082/swagger-ui/index.html -- Order Service
http://localhost:8083/swagger-ui/index.html -- User Service

**Comment by AnuragMishra58** [comment]
Hi @akash-coded 
Uploaded services in zip along with postman collection to test and screenshots
[ecommerce-app-2.zip](https://github.com/akash-coded/spring-framework/files/12826550/ecommerce-app-2.zip)

**Comment by arghyagiri** [comment]
@akash-coded : https://github.com/arghyagiri/microservice-e2/blob/main/e-commerce-microservices-declarative-open-feign/Read.md

**Comment by Rajashree-Ravi** [comment]
@akash-coded 
Please find the attached implementation for feign client... Feign client is used in order service..

[ecommerce-with-feign-client-and-aggregator.zip](https://github.com/akash-coded/spring-framework/files/12827580/ecommerce-with-feign-client-and-aggregator.zip)

**Comment by arulsudarvizhi1** [comment]
Hi Akash,
Please find attached the code base for orderservice and product service with feigh client implementation. The usage of feign client is tested and it worked

Nov 6th to Dec 6th Morning 7:30 batch
Name: Arul
[feignclientusage.zip](https://github.com/akash-coded/spring-framework/files/13330552/feignclientusage.zip)

**Comment by ssagarwala** [comment]
Hello @akash-coded 
this is Fatima. 
I did the homework  the link is https://github.com/ssagarwala/FeignClientExercise.  
Please review and provide feedback. thanks

**Comment by samik2023** [comment]
Hi Aakash,

Attacing here the feign client project. Product and Order service is complete with Feign client connectivity which is working properly. The User service am yet to add full functionality yet.

Name: Samik
Org:  TCS
Batch:  Nov 7th morning 7.30 Batch
[Feign_project.zip](https://github.com/akash-coded/spring-framework/files/13330025/Feign_project.zip)


Thanks
Samik

**Comment by bijumedayil** [comment]
Name Bijumon Habeeb
2255107 
TCS 
[user-product-order.zip](https://github.com/akash-coded/spring-framework/files/13336532/user-product-order.zip)
http://localhost:8081/swagger-ui/index.html#   :Product
http://localhost:8082/swagger-ui/index.html# : Order
http://localhost:8083/swagger-ui/index.html  :User

**Comment by obulix** [comment]
@akash_coded
Name : Dinakaran.AS
https://github.com/obulix/MicSrv-DeclarativePrg

**Comment by pradnyakinge** [comment]
@akash-coded
Name: Pradnya Kinge
https://github.com/pradnyakinge/SpringBoot-Microservices/tree/main/FeignClientDemo

**Comment by AnubhavG1999** [comment]
Hey Akash, dropping my implementation below.

Name: Anubhav Ghosh
Org: TCS
Emp Id: 2286865

[FeignTest.zip](https://github.com/akash-coded/spring-framework/files/13348686/FeignTest.zip)

**Comment by Mycode-Anish** [comment]
Hi Akash,Please find implementation attached.
Name-Anish Sreekumaran Nair'
Tcs
Emp id:375605
[OpenFeignProject.zip](https://github.com/akash-coded/spring-framework/files/13361665/OpenFeignProject.zip)

**Comment by Vamsi-Krishna-Bhagi** [comment]
https://github.com/Vamsi-Krishna-Bhagi/spring-openfeign

**Comment by HarshDPandya** [comment]
Hi @akash-coded , Sharing my code in zip file for **Microservices - Declarative Programming - Spring OpenFeign**
Name: Harsh Pandya
TCS Emp Id: 2066481
[microservice-ecommerce-with-openfeign-HarshPandya-2066481.zip](https://github.com/akash-coded/spring-framework/files/13373392/microservice-ecommerce-with-openfeign-HarshPandya-2066481.zip)

**Comment by ranjanrkms** [comment]
Hi @akash-coded ,
Please find my code for the OpenFiegn.

https://github.com/ranjanrkms/Microservices/tree/main/Ecom_FeignClient

Name- Ranjan Kumar Pandit
Tcs Emp Id- 496856

**Comment by vishalakshi2** [comment]
Hi Akash,
[ecommerce-app.zip](https://github.com/akash-coded/spring-framework/files/13453882/ecommerce-app.zip)

Please find my code for the OpenFeign
Tcs Emp Id-922774

**Comment by jyothi241868** [comment]
Hi Akash,

Please find attached project for Open Feign

[FeignClient.zip](https://github.com/akash-coded/spring-framework/files/13553534/FeignClient.zip)


Thanks,
Jyothi K

**Comment by raghavan-tcs** [comment]
Hi @akash-coded 

Please find attached updated mid-assessment for author-service and book-service using Feign clients.
Batch: Apr-17-2024

Thanks
Raghavan
[mid-assessment_feign_client.zip](https://github.com/akash-coded/spring-framework/files/15071534/mid-assessment_feign_client.zip)

**Comment by Aditi-12092024** [comment]
Hi @akash-coded 
Please find below my work on feign client implementation. 23 APR 2024
[FeignClient.zip](https://github.com/akash-coded/spring-framework/files/15072232/FeignClient.zip)

Thanks

**Comment by umangpathak123** [comment]
Hi @akash-coded

Please find attached Feign client implementation.
[product.zip](https://github.com/akash-coded/spring-framework/files/15080373/product.zip)

Thanks
Umang
